lcls-twincat-motion/Library/POUs/Motion/FB_GantryAutoCoupling.TcPOU

```diff
@@ -0,0 +1,89 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4020.14">
+  <POU Name="FB_GantryAutoCoupling" Id="{532dfcaf-6187-4140-aaad-e20a9bbe6b43}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_GantryAutoCoupling
+VAR_INPUT
+	fGantryTol : LREAL;
+END_VAR
+VAR_OUTPUT
+	bGantryAlreadyCoupled : BOOL;
+END_VAR
+VAR_IN_OUT
+	Master : DUT_MotionStage;
+	MasterEnc : ST_RenishawAbsEnc;
+	Slave : DUT_MotionStage;
+	SlaveEnc : ST_RenishawAbsEnc;
+	bExecuteCouple : BOOL;
+	bExecuteDecouple : BOOL;
+END_VAR
+VAR
+	gantry_diff_limit : FB_GantryDiffVirtualLimitSwitch;
+	couple : MC_GEARIN;
+	decouple : MC_GEAROUT;
+
+
+	bInitComplete : BOOL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Designate Master and SLave Axes
+Master.bGantryAxis := TRUE;
+Slave.bGantryAxis := TRUE;
+
+Master.fGantryTol := fGantryTol;
+Slave.fGantryTol := Master.fGantryTol;
+
+// Activate Gantry Virtual Limit Switch
+gantry_diff_limit(Penc:=MasterEnc, SEnc:=SlaveEnc, GantDiffTol:=LREAL_TO_LINT(Master.fGantryTol),
+				  PLimFwd=>Master.bLimGntFwd, PLimBwd=>Master.bLimGntBwd,
+				  SLimFwd=>Slave.bLimGntFwd, SLimBwd=>Slave.bLimGntBwd);
+
+// Coupling Status Bit
+bGantryAlreadyCoupled := Master.Axis.NcToPlc.CoupleState=1 AND Slave.Axis.NcToPlc.CoupleState=3;
+
+// Make master and slave axes respect each other's limits
+IF bGantryAlreadyCoupled THEN
+	// Fwd limits
+	Master.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Slave.bLimFwd;
+	Slave.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Master.bLimFwd;
+	// Bwd limits
+	Master.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Slave.bLimBwd;
+	Slave.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Master.bLimBwd;
+END_IF
+
+// Coupling states
+// Auto-coupling at init and auto-reset of coupling boolean
+bExecuteCouple S= NOT bInitComplete;
+
+bExecuteCouple R= couple.Busy OR bGantryAlreadyCoupled;
```

Comment:
What happens if `couple.Busy` is `TRUE` and we reset this, then enter the active `couple` function block with the `Execute` arg set to `FALSE`? Does this stop the coupling in progress? Or does it do nothing? Or does this never happen because the coupling happens in one cycle?

Reply:
So I'm not entirely sure and trying to figure out a way to step through the cycles and test this but here are my initial thoughts:
1. Coupling occurs on the second cycle, so what you're saying could potentially happen.
2. If we enter the active `couple` function block with `bExecuteCouple = FALSE`, it won't stop coupling if it has already occurred, that was why I added the `decouple` function block to provide this functionality.
3. Since coupling didn't execute on this cycle, it seems to me that `bInitComplete` would remain False which then would set `bExecuteCouple` to True on the next cycle and it would just take an extra cycle to complete coupling.

I need to try and step through this though as I could be overlooking something

Reply:
I just tried this by setting breakpoints from the `bExecuteCouple S= ` line until the end and I don't think this will be an issue. I found that `couple.Busy` stays False until after the `couple` function block's execution begins and then goes to True. Even if someone were to reset this to False before coupling finishes, this doesn't stop the `couple` function block that's currently getting executed, and then after it completes `couple.Busy` goes to False anyway so the desired logic continues executing.